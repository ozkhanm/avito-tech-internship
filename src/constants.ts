export enum PAGE {
    ARTICLE = "ARTICLE_PAGE",
    MAIN = "MAIN_PAGE",
};

export const REFRESH_TIMEOUT = 60000;

export const NOT_FOUND_ERROR_MESSAGE = "Article not found";

export const DATA_LOADING_MESSAGE = "Data loading...";

export const REQUEST_URL = "https://hacker-news.firebaseio.com/v0/";

export const REQUEST_TIMEOUT = 10000;

export const REQUEST_ITEMS = 100;