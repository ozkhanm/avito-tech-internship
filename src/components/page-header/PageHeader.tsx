import { useEffect, useState } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";

import { ActionCreator } from "../../reducer/action-creator";
import { AppDispatch, Operation, RootState } from "../../reducer/reducer";
import { PAGE } from "../../constants";

interface PageHeaderProps {
    dropActiveArticle: () => void,
    getArticles: () => void,
    changeCommentsLoadingStatus: (status: boolean) => void,
    activeArticleId: number,
    getActiveArticle: (id: number) => void,
    isCommentLoaded: boolean,
    refreshStatus: boolean,
    page: string,
    errorMessage: string,
    isActiveArticleLoaded: boolean,
    changeRefreshStatus: (status: boolean) => void,
}

const PageHeader: React.FC<PageHeaderProps> = (props) => {
    const { dropActiveArticle, getArticles, changeCommentsLoadingStatus, 
        activeArticleId, getActiveArticle, isCommentLoaded, refreshStatus, 
        page, errorMessage, isActiveArticleLoaded, changeRefreshStatus } = props;
    const [refreshButtonToggled, setRefreshButtonToggleStatus] = useState(false);

    useEffect(() => {
        let refreshButtonToggleStatusTimeout: undefined | ReturnType<typeof setTimeout>;

        if (refreshButtonToggled) {
            refreshButtonToggleStatusTimeout = setTimeout(() => {
                setRefreshButtonToggleStatus(false);
                changeRefreshStatus(false);
            }, 3000);
        }

        return () => {
            refreshButtonToggleStatusTimeout ? clearTimeout(refreshButtonToggleStatusTimeout) : undefined;
        };
    });

    const renderHeaderControls = () => {
        const getRefreshButtonAnimationClass = () => {
            let conditions;

            if (page === PAGE.MAIN) {
                conditions = refreshButtonToggled || refreshStatus;
            } else {
                conditions = (refreshButtonToggled && !isCommentLoaded) || refreshStatus;
            }

            return conditions ? "header-controls-refresh-button-animation" : "";
        };

        const getRefreshButtonClassName = () => {
            return `header-controls-button header-controls-refresh-button ${getRefreshButtonAnimationClass()}`;
        };

        const refreshButtonClassName = getRefreshButtonClassName();

        if (page === PAGE.MAIN) {
            return (
                <div className="header-controls">
                    <a className={refreshButtonClassName} onClick={() => {
                        setRefreshButtonToggleStatus(true);
                        getArticles();
                    }}/>
                </div>
            );
        } else {
            return (
                <div className="header-controls">
                    <a className={refreshButtonClassName} onClick={() => {
                        setRefreshButtonToggleStatus(true);
                        changeCommentsLoadingStatus(false);
                        getActiveArticle(activeArticleId);
                    }}/>
                    <Link className="header-controls-button header-controls-return-button" to={"/"} onClick={() => {
                        dropActiveArticle();
                    }}/>
                </div>
            );
        }
    };

    return (
        <>
            <div className="header-container">
                { !!errorMessage || !isActiveArticleLoaded ? null : renderHeaderControls() }
                <p className="header">Hacker News</p>
            </div>
        </>
    );
};

const mapStateToProps = (state: RootState) => ({
    activeArticleId: state.activeArticleId,
    isCommentLoaded: state.isCommentLoaded,
    refreshStatus: state.refreshStatus,
    errorMessage: state.errorMessage,
    isActiveArticleLoaded: state.isActiveArticleLoaded,
});

const mapDispatchToProps = (dispatch: AppDispatch) => ({
    dropActiveArticle() {
        dispatch(ActionCreator.dropActiveArticle());
    },
    getArticles() {
        dispatch(Operation.getArticles());
    },
    changeCommentsLoadingStatus(status: boolean) {
        dispatch(ActionCreator.changeCommentsLoadingStatus(status));
    },
    getActiveArticle(articleId: number) {
        dispatch(Operation.getActiveArticle(articleId));
    },
    changeRefreshStatus(status: boolean) {
        dispatch(ActionCreator.changeRefreshStatus(status));
    },
});

export { PageHeader };
export default connect(mapStateToProps, mapDispatchToProps)(PageHeader);