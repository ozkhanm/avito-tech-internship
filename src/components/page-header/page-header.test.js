import renderer from "react-test-renderer";
import { Router } from "react-router-dom";

import { PageHeader } from "./PageHeader";

import history from "../../history";
import { PAGE } from "../../constants";

it("Page header renders correctly on article page screen", () => {
    const tree = renderer
        .create(
            <Router history={history}>
                <PageHeader page={PAGE.ARTICLE}/>
            </Router>)
        .toJSON();
  
    expect(tree).toMatchSnapshot();
});

it("Page header renders correctly on main page screen", () => {
    const tree = renderer
        .create(
            <Router history={history}>
                <PageHeader page={PAGE.MAIN}/>
            </Router>)
        .toJSON();
  
    expect(tree).toMatchSnapshot();
});