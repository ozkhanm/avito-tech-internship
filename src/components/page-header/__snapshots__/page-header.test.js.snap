// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page header renders correctly on article page screen 1`] = `
<div
  className="header-container"
>
  <p
    className="header"
  >
    Hacker News
  </p>
</div>
`;

exports[`Page header renders correctly on main page screen 1`] = `
<div
  className="header-container"
>
  <p
    className="header"
  >
    Hacker News
  </p>
</div>
`;
